packages/fuel-indexer-plugin/src/find.rs

```diff
@@ -0,0 +1,298 @@
+use fuel_indexer_types::scalar::{Boolean, UID};
+use sqlparser::ast as sql;
+
+/// Represents `filter` and `order_by` parts of the `SELECT object from {table}
+/// WHERE {filter} {order_by}` statement that is assembled by the indexer to
+/// fetch an object from the database. The table name is not available to the
+/// plugin and thus only a part of the statment is generated there. The indexer
+/// maps the TYPE_ID to the tale name and assemles the full statemnt.
+pub struct QueryFragment<T> {
+    filter: Filter<T>,
+    order_by: Option<sql::OrderByExpr>,
+}
+
+/// Convert `QueryFragment` to `String`. `SELECT * from table_name` is lated
+/// added by the Fuel indexer to generate the entire query.
+impl<T> std::fmt::Display for QueryFragment<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.filter)?;
+        if let Some(ref order_by) = self.order_by {
+            write!(f, " ORDER BY {}", order_by)?;
+        }
+        Ok(())
+    }
+}
+
+/// Automatic lifting of `Filter` into `QueryFragment` leaving `ORDER BY`
+/// unspecified.
+impl<T> From<Filter<T>> for QueryFragment<T> {
+    fn from(filter: Filter<T>) -> Self {
+        QueryFragment {
+            filter,
+            order_by: None,
+        }
+    }
+}
+
+/// Represents a WHERE clause of the SQL statement. Multiple `Filter`s can be
+/// joined with `and` and `or` and also ordered, at which point they become
+/// `QueryFragment`s.
+pub struct Filter<T> {
+    filter: sql::Expr,
+    phantom: std::marker::PhantomData<T>,
+}
+
+impl<T> std::fmt::Display for Filter<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.filter)
+    }
+}
+
+impl<T> Filter<T> {
+    fn new(filter: sql::Expr) -> Self {
+        Self {
+            filter,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn and(self, right: Filter<T>) -> Filter<T> {
+        let filter = sql::Expr::BinaryOp {
+            left: Box::new(self.filter),
+            op: sql::BinaryOperator::And,
+            right: Box::new(right.filter),
+        };
+        Filter {
+            filter,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn or(self, right: Filter<T>) -> Filter<T> {
+        let filter = sql::Expr::BinaryOp {
+            left: Box::new(self.filter),
+            op: sql::BinaryOperator::Or,
+            right: Box::new(right.filter),
+        };
+        Filter {
+            filter,
+            phantom: std::marker::PhantomData,
+        }
+    }
+
+    pub fn order_by_asc<F>(self, f: Field<T, F>) -> QueryFragment<T> {
```

Comment:
- So for this,  I think the more ORM way to do `order by` is to follow the SQL syntax.
- In SQL, order by is:

```
SELECT foo
FROM bar
ORDER BY {field} {direction}
```

- Where `{direction}` is `ASC/DESC` and `{field}` is obviously the field you're ordering by.
- I think functionally what you have is roughly the same as that ☝🏼 
- But syntactically, it's not "SQL"
- Some links
  - [`order_by`](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.query.QuerySet.order_by) in Django
    - You first declare your field, then you declare how to order it
  - [`order_by`](https://sequelize.org/docs/v6/core-concepts/model-querying-basics/#ordering-and-grouping) in Sequelize
    - Again, you first declare what to order, then how to order it 
  - Finally, [`order_by`](https://docs.diesel.rs/2.0.x/diesel/prelude/trait.QueryDsl.html#method.order_by) in Diesel 
    - Again, declaring what to order first, then how to order it

So I think a small change to something like:

```rust
Struct::find(Struct::field().eq(5).order_by(Struct::field().desc()); // like diesel
```
OR
```rust
Struct::find(Struct::field().eq(5).order_by(Struct::field(), Order::DESC));
```